### target idea:

At date T, using only information available at T,
what happened to the stock after T?

### import lib

In [1]:
from pathlib import Path

import numpy as np
import pandas as pd

### path to files and load them

In [2]:
def find_project_root(start_path: Path) -> Path:
    """
    Walk upward from the current notebook folder until we find the project root.
    Project root is identified by config.yaml or README.md.
    """
    start_path = start_path.resolve()

    for path in [start_path] + list(start_path.parents):
        if (path / "config.yaml").exists() or (path / "README.md").exists():
            return path

    raise FileNotFoundError("Project root not found. Expected config.yaml or README.md.")


PROJECT_ROOT = find_project_root(Path.cwd())

INTERIM_DIR = PROJECT_ROOT / "data" / "interim"
FEATURE_DIR = INTERIM_DIR / "features"
TARGET_DIR = INTERIM_DIR / "targets"

TARGET_DIR.mkdir(parents=True, exist_ok=True)

FEATURE_INPUT_PARQUET = FEATURE_DIR / "stock_features_v1.parquet"
FEATURE_INPUT_CSV = FEATURE_DIR / "stock_features_v1.csv"

TARGET_OUTPUT_PARQUET = TARGET_DIR / "stock_features_with_targets_v1.parquet"
TARGET_OUTPUT_CSV = TARGET_DIR / "stock_features_with_targets_v1.csv"

TARGET_QUALITY_PATH = TARGET_DIR / "target_quality_summary_v1.csv"
TARGET_DICTIONARY_PATH = TARGET_DIR / "target_dictionary_v1.csv"

print("Project root:", PROJECT_ROOT)
print("Feature parquet:", FEATURE_INPUT_PARQUET)
print("Target dir:", TARGET_DIR)

if FEATURE_INPUT_PARQUET.exists():
    stock_features = pd.read_parquet(FEATURE_INPUT_PARQUET)
    print("Loaded from parquet")
else:
    stock_features = pd.read_csv(FEATURE_INPUT_CSV, parse_dates=["date"])
    print("Loaded from CSV")

stock_features["date"] = pd.to_datetime(stock_features["date"])

stock_features = (
    stock_features
    .sort_values(["yf_ticker", "date"])
    .reset_index(drop=True)
)
# to display all the columns in the DataFrame without truncation
pd.set_option("display.max_columns", None)
print("Stock features shape:", stock_features.shape)
print("Date range:", stock_features["date"].min(), "to", stock_features["date"].max())
print("Unique stocks:", stock_features["yf_ticker"].nunique())

display(stock_features.sample(5))

Project root: E:\Projects\marketguard-india
Feature parquet: E:\Projects\marketguard-india\data\interim\features\stock_features_v1.parquet
Target dir: E:\Projects\marketguard-india\data\interim\targets
Loaded from parquet
Stock features shape: (357370, 142)
Date range: 2010-01-04 00:00:00 to 2026-07-14 00:00:00
Unique stocks: 100


,asset_type,symbol,yf_ticker,company_name,industry,date,adj_close,close,dividends,high,low,open,repaired?,stock_splits,volume,source_file,history_days_available,prev_close,prev_adj_close,return_1d,return_5d,return_20d,return_60d,return_120d,log_return_1d,raw_return_1d,daily_volatility_20d,daily_volatility_60d,annualized_volatility_20d,annualized_volatility_60d,ma_20,adj_close_to_ma_20,ma_50,adj_close_to_ma_50,ma_100,adj_close_to_ma_100,ma_200,adj_close_to_ma_200,rolling_high_60d,rolling_high_252d,drawdown_from_60d_high,drawdown_from_252d_high,rsi_14,above_ma_20,above_ma_50,above_ma_200,ma_20_above_ma_50,ma_50_above_ma_200,true_range,atr_14,atr_14_pct,intraday_range_pct,open_gap_pct,close_position_in_day_range,volume_ma_20,volume_ma_60,volume_ratio_20d,volume_ratio_60d,turnover,turnover_ma_20,turnover_ratio_20d,large_raw_move_flag,stock_acceptance_status,stock_history_rows,stock_history_start_date,stock_history_end_date,is_research_universe,is_limited_history,large_move_reason,review_status,large_move_review_flag,manual_review_large_move_flag,keep_large_move_flag,nifty50_close,nifty50_return_1d,nifty50_return_5d,nifty50_return_20d,nifty50_return_60d,nifty50_daily_volatility_20d,nifty50_daily_volatility_60d,nifty50_annualized_volatility_20d,nifty50_annualized_volatility_60d,nifty50_close_to_ma_20,nifty50_close_to_ma_50,nifty50_close_to_ma_200,nifty100_close,nifty100_return_1d,nifty100_return_5d,nifty100_return_20d,nifty100_return_60d,nifty100_daily_volatility_20d,nifty100_daily_volatility_60d,nifty100_annualized_volatility_20d,nifty100_annualized_volatility_60d,nifty100_close_to_ma_20,nifty100_close_to_ma_50,nifty100_close_to_ma_200,indiavix_close,indiavix_return_1d,indiavix_return_5d,indiavix_return_20d,indiavix_return_60d,indiavix_daily_volatility_20d,indiavix_daily_volatility_60d,indiavix_annualized_volatility_20d,indiavix_annualized_volatility_60d,indiavix_close_to_ma_20,indiavix_close_to_ma_50,indiavix_close_to_ma_200,sector_proxy_ticker,sector_proxy_name,sector_close,sector_return_1d,sector_return_5d,sector_return_20d,sector_return_60d,sector_daily_volatility_20d,sector_daily_volatility_60d,sector_annualized_volatility_20d,sector_annualized_volatility_60d,sector_close_to_ma_20,sector_close_to_ma_50,sector_close_to_ma_200,relative_return_20d_vs_nifty50,relative_return_60d_vs_nifty50,relative_return_20d_vs_nifty100,relative_return_60d_vs_nifty100,relative_return_20d_vs_sector,relative_return_60d_vs_sector,volatility_ratio_vs_nifty50_20d,volatility_ratio_vs_nifty100_20d,volatility_ratio_vs_sector_20d,nifty50_uptrend_flag,nifty100_uptrend_flag,vix_above_20_flag,vix_above_25_flag,vix_20d_change,base_features_ready,market_features_ready,sector_features_ready,model_ready_v1,model_ready_v1_clean
302105,stock,TCS,TCS.NS,Tata Consultancy Services Ltd.,Information Technology,2011-06-17,395.607941,554.599976,0.0,578.500000,552.000000,577.500000,False,0.0,3788146,TCS_NS.csv,365,575.575012,410.569946,-0.036442,-0.067272,-0.049553,0.044203,-0.019667,-0.037123,-0.036442,0.014412,0.017246,0.228781,0.273772,412.783623,-0.041609,412.578893,-0.041134,406.262721,-0.026226,386.038714,0.024788,439.205963,439.205963,-0.099266,-0.099266,38.026280,0,0,1,1,1,26.500000,12.469644,0.022484,0.047782,0.003344,0.098112,2141523.60,2.893033e+06,1.768902,1.309403,2.100906e+09,1.244761e+09,1.687798,0,OK,4082,2010-01-04,2026-07-14,1,0,NaN,NaN,0,0,0,5366.399902,-0.005624,-0.021765,-0.011367,0.000308,0.009395,0.010375,0.149144,0.164694,-0.019417,-0.042099,-0.067087,5307.950195,-0.005844,-0.021486,-0.017492,-0.001402,0.008229,0.009763,0.130631,0.154981,-0.019372,-0.037315,-0.065734,20.270000,0.031027,0.093312,0.042159,-0.089398,0.047320,0.039722,0.751183,0.630567,0.085408,0.009402,-0.038468,^CNXIT,nifty_it,6357.100098,-0.019080,-0.044541,-0.039045,-0.027066,0.008820,0.014017,0.140013,0.222509,-0.032264,-0.048945,-0.065261,-0.038186,0.043896,-0.032061,0.045605,-0.010507,0.071270,1.533953,1.751357,1.633992,0,0,1,0,0.042159,1,1,1,1,1
210908,stock,M&M,M&M.NS,Mahindra

### Target engineering config and source-column check

For targets, we will use adj_close, not raw close, because it is adjusted for corporate actions.

In [3]:
# Target horizons:
# 5 trading days  ≈ 1 week
# 20 trading days ≈ 1 month
# 60 trading days ≈ 3 months
TARGET_HORIZONS = [5, 20, 60]

# Main horizon for v1 model/risk labels
MAIN_TARGET_HORIZON = 20

# Downside / upside thresholds for classification-style labels
# These are not trading rules yet; they are target labels for analysis/model training.
DOWNSIDE_RISK_THRESHOLD_20D = -0.05   # -5% or worse over next 20 trading days
BIG_DOWNSIDE_THRESHOLD_20D = -0.10    # -10% or worse over next 20 trading days
UPSIDE_THRESHOLD_20D = 0.05           # +5% or better over next 20 trading days

required_source_cols = [
    "date",
    "symbol",
    "yf_ticker",
    "adj_close",
    "close",
    "model_ready_v1",
    "model_ready_v1_clean",
    "stock_acceptance_status",
    "manual_review_large_move_flag",
]

missing_source_cols = [
    col for col in required_source_cols
    if col not in stock_features.columns
]

print("Missing source columns:", missing_source_cols)

if missing_source_cols:
    raise ValueError(f"Missing source columns: {missing_source_cols}")

print("Target horizons:", TARGET_HORIZONS)
print("Main target horizon:", MAIN_TARGET_HORIZON)

print("\nAdj close missing values:")
print(stock_features["adj_close"].isna().sum())

print("\nModel-ready rows available before target creation:")
display(stock_features["model_ready_v1"].value_counts())

display(
    stock_features[
        [
            "date",
            "symbol",
            "yf_ticker",
            "adj_close",
            "close",
            "model_ready_v1",
            "model_ready_v1_clean",
            "stock_acceptance_status",
        ]
    ]
    .sample(20)
)

Missing source columns: []
Target horizons: [5, 20, 60]
Main target horizon: 20

Adj close missing values:
0

Model-ready rows available before target creation:


model_ready_v1
1    309073
0     48297
Name: count, dtype: int64

,date,symbol,yf_ticker,adj_close,close,model_ready_v1,model_ready_v1_clean,stock_acceptance_status
65262,2023-11-01,BHARTIARTL,BHARTIARTL.NS,900.856689,913.450012,1,1,OK
232475,2020-02-03,NTPC,NTPC.NS,85.924484,110.250000,1,1,OK
242221,2026-06-24,PFC,PFC.NS,437.100006,437.100006,1,1,OK
165955,2010-03-22,INDHOTEL,INDHOTEL.NS,83.815758,91.702065,0,0,OK
144882,2017-12-08,HDFCBANK,HDFCBANK.NS,422.412781,460.100006,1,1,OK
105664,2022-12-30,DIVISLAB,DIVISLAB.NS,3349.953613,3413.199951,1,1,OK
326412,2010-01-19,TVSMOTOR,TVSMOTOR.NS,32.010620,38.674999,0,0,OK
177169,2011-11-18,IOC,IOC.NS,21.335295,45.325001,0,0,OK
139995,2022-08-12,HCLTECH,HCLTECH.NS,820.955872,956.400024,1,1,OK
44013,2020-06-18,BAJAJFINSV,BAJAJFINSV.NS,541.075012,542.755005,1,1,OK


### Future return targets

Now create the basic future-return targets.

#### From today, what was the stock return after 5 / 20 / 60 trading days?

In [4]:
stock_targets = stock_features.copy()

# Create future adjusted-close prices and future returns.
# Example:
# future_return_20d at date T =
#     adjusted close 20 trading days after T / adjusted close at T - 1
for horizon in TARGET_HORIZONS:
    future_price_col = f"future_adj_close_{horizon}d"
    future_return_col = f"future_return_{horizon}d"
    target_available_col = f"target_available_{horizon}d"

    stock_targets[future_price_col] = (
        stock_targets
        .groupby("yf_ticker")["adj_close"]
        .shift(-horizon)
    )

    stock_targets[future_return_col] = (
        stock_targets[future_price_col] / stock_targets["adj_close"] - 1
    )

    # 1 means the future target is available.
    # Last N rows per stock will be 0 because we do not know the future after dataset end.
    stock_targets[target_available_col] = (
        stock_targets[future_return_col].notna()
    ).astype(int)

print("Stock targets shape:", stock_targets.shape)

for horizon in TARGET_HORIZONS:
    print(f"\nFuture return {horizon}d available:")
    display(stock_targets[f"target_available_{horizon}d"].value_counts())

display(
    stock_targets[
        [
            "date",
            "symbol",
            "yf_ticker",
            "adj_close",
            "future_adj_close_5d",
            "future_return_5d",
            "future_adj_close_20d",
            "future_return_20d",
            "future_adj_close_60d",
            "future_return_60d",
        ]
    ]
    .dropna()
    .sample(30, random_state=42)
)

Stock targets shape: (357370, 151)

Future return 5d available:


target_available_5d
1    356870
0       500
Name: count, dtype: int64


Future return 20d available:


target_available_20d
1    355370
0      2000
Name: count, dtype: int64


Future return 60d available:


target_available_60d
1    351370
0      6000
Name: count, dtype: int64

,date,symbol,yf_ticker,adj_close,future_adj_close_5d,future_return_5d,future_adj_close_20d,future_return_20d,future_adj_close_60d,future_return_60d
223111,2015-03-19,MUTHOOTFIN,MUTHOOTFIN.NS,157.090225,157.090225,0.000000,153.837585,-0.020706,149.179504,-0.050358
296753,2022-11-16,TATAPOWER,TATAPOWER.NS,218.849609,215.146912,-0.016919,216.998260,-0.008459,200.238632,-0.085040
216388,2025-06-24,MAXHEALTH,MAXHEALTH.NS,1194.795532,1275.957397,0.067930,1216.637939,0.018281,1171.318359,-0.019650
188732,2020-03-18,JINDALSTEL,JINDALSTEL.NS,99.839005,90.915131,-0.089383,81.744720,-0.181235,143.127243,0.433580
66766,2013-05-21,BOSCHLTD,BOSCHLTD.NS,8193.377930,8082.816895,-0.013494,8072.224609,-0.014787,7613.045410,-0.070829
356229,2021-12-06,ZYDUSLIFE,ZYDUSLIFE.NS,438.994507,450.878876,0.027072,463.005768,0.054696,344.453003,-0.215359
111392,2020-05-14,DMART,DMART.NS,2324.399902,2470.250000,0.062747,2397.800049,0.031578,2138.399902,-0.080021
137083,2010-10-21,HCLTECH,HCLTECH.NS,71.564003,69.745575,-0.025410,66.321159,-0.073261,83.060776,0.160650
43779,2019-07-03,BAJAJFINSV,BAJAJFINSV.NS,839.689880,762.772217,-0.091602,707.191040,-0.157795,842.743896,0.003637
48988,2024-01-19,BAJAJHLDNG,BAJAJHLDNG.NS,8240.452148,8025.366211,-0.026101,8575.699219,0.040683,7879.188477,-0.043840


### Future downside / upside path targets

Now we create future downside and upside targets.

These are more useful for MarketGuard than only final return.

future_return_20d = where price ended after 20 days.

future_min_return_20d = worst drop seen anytime during next 20 days.

future_max_return_20d = best upside seen anytime during next 20 days.

In [5]:
# Rebuild stock_targets cleanly from stock_features
# This avoids using the partially broken result from the previous failed cell.
stock_targets = stock_features.copy()

# Recreate future adjusted-close prices and future returns
for horizon in TARGET_HORIZONS:
    future_price_col = f"future_adj_close_{horizon}d"
    future_return_col = f"future_return_{horizon}d"
    target_available_col = f"target_available_{horizon}d"

    stock_targets[future_price_col] = (
        stock_targets
        .groupby("yf_ticker")["adj_close"]
        .shift(-horizon)
    )

    stock_targets[future_return_col] = (
        stock_targets[future_price_col] / stock_targets["adj_close"] - 1
    )

    stock_targets[target_available_col] = (
        stock_targets[future_return_col].notna()
    ).astype(int)


def add_future_path_targets(g: pd.DataFrame, horizons: list[int]) -> pd.DataFrame:
    """
    Add future path-based targets for one stock.

    For each date T and horizon H:
    - future_min_return_Hd = lowest adjusted close from T+1 to T+H, relative to T
    - future_max_return_Hd = highest adjusted close from T+1 to T+H, relative to T
    """
    g = g.sort_values("date").copy()

    for horizon in horizons:
        future_price_cols = []

        for step in range(1, horizon + 1):
            col = f"_future_adj_close_step_{step}"
            g[col] = g["adj_close"].shift(-step)
            future_price_cols.append(col)

        full_future_window_available = g[future_price_cols].notna().all(axis=1)

        min_price_col = f"future_min_adj_close_{horizon}d"
        max_price_col = f"future_max_adj_close_{horizon}d"
        min_return_col = f"future_min_return_{horizon}d"
        max_return_col = f"future_max_return_{horizon}d"
        path_available_col = f"path_target_available_{horizon}d"

        g[min_price_col] = g[future_price_cols].min(axis=1)
        g[max_price_col] = g[future_price_cols].max(axis=1)

        g[min_return_col] = g[min_price_col] / g["adj_close"] - 1
        g[max_return_col] = g[max_price_col] / g["adj_close"] - 1

        g.loc[
            ~full_future_window_available,
            [min_price_col, max_price_col, min_return_col, max_return_col]
        ] = np.nan

        g[path_available_col] = full_future_window_available.astype(int)

        g = g.drop(columns=future_price_cols)

    return g


target_frames = []

for ticker, g in stock_targets.groupby("yf_ticker", sort=False):
    gf = add_future_path_targets(g, TARGET_HORIZONS)
    target_frames.append(gf)

stock_targets = (
    pd.concat(target_frames, ignore_index=True)
    .sort_values(["yf_ticker", "date"])
    .reset_index(drop=True)
)

print("Stock targets shape:", stock_targets.shape)
print("yf_ticker exists:", "yf_ticker" in stock_targets.columns)

for horizon in TARGET_HORIZONS:
    print(f"\nFuture path target {horizon}d available:")
    display(stock_targets[f"path_target_available_{horizon}d"].value_counts())

display(
    stock_targets[
        [
            "date",
            "symbol",
            "yf_ticker",
            "adj_close",
            "future_return_20d",
            "future_min_return_20d",
            "future_max_return_20d",
            "target_available_20d",
            "path_target_available_20d",
        ]
    ]
    .dropna()
    .sample(30, random_state=42)
)

Stock targets shape: (357370, 166)
yf_ticker exists: True

Future path target 5d available:


path_target_available_5d
1    356870
0       500
Name: count, dtype: int64


Future path target 20d available:


path_target_available_20d
1    355370
0      2000
Name: count, dtype: int64


Future path target 60d available:


path_target_available_60d
1    351370
0      6000
Name: count, dtype: int64

,date,symbol,yf_ticker,adj_close,future_return_20d,future_min_return_20d,future_max_return_20d,target_available_20d,path_target_available_20d
247835,2016-02-25,PNB,PNB.NS,64.692307,0.142160,-0.010453,0.204181,1,1
298123,2011-11-16,TATASTEEL,TATASTEEL.NS,26.242041,-0.075625,-0.076364,0.032393,1,1
191332,2011-04-07,JSWSTEEL,JSWSTEEL.NS,83.956139,-0.111871,-0.141437,-0.002946,1,1
311652,2017-02-02,TITAN,TITAN.NS,375.441833,0.129848,-0.024531,0.174672,1,1
229433,2024-04-18,NESTLEIND,NESTLEIND.NS,1200.770020,-0.006924,-0.010091,0.040649,1,1
204788,2017-06-23,LTM,LTM.NS,701.637207,-0.052488,-0.052488,-0.010659,1,1
54142,2011-11-04,BANKBARODA,BANKBARODA.NS,124.497688,-0.090892,-0.158156,0.004971,1,1
332683,2018-12-04,ULTRACEMCO,ULTRACEMCO.NS,3899.454590,-0.006091,-0.063042,0.013373,1,1
188073,2017-07-13,JINDALSTEL,JINDALSTEL.NS,136.520584,-0.109426,-0.109426,0.144095,1,1
280244,2022-11-23,SIEMENS,SIEMENS.NS,1604.300781,0.041412,-0.002166,0.096976,1,1


### 20-day target labels

In [6]:
target_horizon = MAIN_TARGET_HORIZON

target_available_mask = (
    (stock_targets[f"target_available_{target_horizon}d"] == 1)
    & (stock_targets[f"path_target_available_{target_horizon}d"] == 1)
)

# Rows usable for target modelling:
# Features must be ready AND the future 20-day target must be available.
stock_targets["target_ready_v1_20d"] = (
    (stock_targets["model_ready_v1"] == 1)
    & target_available_mask
).astype(int)

stock_targets["target_ready_v1_clean_20d"] = (
    (stock_targets["model_ready_v1_clean"] == 1)
    & target_available_mask
).astype(int)


def make_target_flag(condition: pd.Series, available_mask: pd.Series) -> pd.Series:
    """
    Create nullable target flag.
    - 1 / 0 when target is available
    - NA when future target is not available
    """
    out = pd.Series(pd.NA, index=condition.index, dtype="Int64")
    out.loc[available_mask] = condition.loc[available_mask].astype(int)
    return out


# Final 20-day return direction:
# 1 means stock ended positive after 20 trading days.
stock_targets["target_positive_return_20d"] = make_target_flag(
    stock_targets["future_return_20d"] > 0,
    target_available_mask
)

# Downside risk label:
# 1 means stock fell at least 5% at some point within the next 20 trading days.
stock_targets["target_downside_5pct_20d"] = make_target_flag(
    stock_targets["future_min_return_20d"] <= DOWNSIDE_RISK_THRESHOLD_20D,
    target_available_mask
)

# Big downside risk label:
# 1 means stock fell at least 10% at some point within the next 20 trading days.
stock_targets["target_big_downside_10pct_20d"] = make_target_flag(
    stock_targets["future_min_return_20d"] <= BIG_DOWNSIDE_THRESHOLD_20D,
    target_available_mask
)

# Upside opportunity label:
# 1 means stock rose at least 5% at some point within the next 20 trading days.
stock_targets["target_upside_5pct_20d"] = make_target_flag(
    stock_targets["future_max_return_20d"] >= UPSIDE_THRESHOLD_20D,
    target_available_mask
)

# Simple 3-class forward return bucket.
# This uses final 20-day return, not path high/low.
# -1 = weak, 0 = neutral, 1 = strong
stock_targets["target_return_bucket_20d"] = pd.Series(
    pd.NA,
    index=stock_targets.index,
    dtype="Int64"
)

stock_targets.loc[
    target_available_mask & (stock_targets["future_return_20d"] <= -0.03),
    "target_return_bucket_20d"
] = -1

stock_targets.loc[
    target_available_mask
    & (stock_targets["future_return_20d"] > -0.03)
    & (stock_targets["future_return_20d"] < 0.03),
    "target_return_bucket_20d"
] = 0

stock_targets.loc[
    target_available_mask & (stock_targets["future_return_20d"] >= 0.03),
    "target_return_bucket_20d"
] = 1


print("Stock targets shape:", stock_targets.shape)

print("\nTarget-ready v1 20d:")
display(stock_targets["target_ready_v1_20d"].value_counts())

print("\nTarget-ready v1 clean 20d:")
display(stock_targets["target_ready_v1_clean_20d"].value_counts())

target_label_cols = [
    "target_positive_return_20d",
    "target_downside_5pct_20d",
    "target_big_downside_10pct_20d",
    "target_upside_5pct_20d",
    "target_return_bucket_20d",
]

for col in target_label_cols:
    print(f"\n{col}:")
    display(stock_targets[col].value_counts(dropna=False))

display(
    stock_targets[
        [
            "date",
            "symbol",
            "future_return_20d",
            "future_min_return_20d",
            "future_max_return_20d",
            "target_positive_return_20d",
            "target_downside_5pct_20d",
            "target_big_downside_10pct_20d",
            "target_upside_5pct_20d",
            "target_return_bucket_20d",
            "target_ready_v1_20d",
        ]
    ]
    .dropna()
    .sample(30, random_state=42)
)

Stock targets shape: (357370, 173)

Target-ready v1 20d:


target_ready_v1_20d
1    307453
0     49917
Name: count, dtype: int64


Target-ready v1 clean 20d:


target_ready_v1_clean_20d
1    307418
0     49952
Name: count, dtype: int64


target_positive_return_20d:


target_positive_return_20d
1       200618
0       154752
<NA>      2000
Name: count, dtype: Int64


target_downside_5pct_20d:


target_downside_5pct_20d
0       218952
1       136418
<NA>      2000
Name: count, dtype: Int64


target_big_downside_10pct_20d:


target_big_downside_10pct_20d
0       303823
1        51547
<NA>      2000
Name: count, dtype: Int64


target_upside_5pct_20d:


target_upside_5pct_20d
0       178122
1       177248
<NA>      2000
Name: count, dtype: Int64


target_return_bucket_20d:


target_return_bucket_20d
1       147378
0       105243
-1      102749
<NA>      2000
Name: count, dtype: Int64

,date,symbol,future_return_20d,future_min_return_20d,future_max_return_20d,target_positive_return_20d,target_downside_5pct_20d,target_big_downside_10pct_20d,target_upside_5pct_20d,target_return_bucket_20d,target_ready_v1_20d
247835,2016-02-25,PNB,0.142160,-0.010453,0.204181,1,0,0,1,1,1
298123,2011-11-16,TATASTEEL,-0.075625,-0.076364,0.032393,0,1,0,0,-1,0
191332,2011-04-07,JSWSTEEL,-0.111871,-0.141437,-0.002946,0,1,1,0,-1,0
311652,2017-02-02,TITAN,0.129848,-0.024531,0.174672,1,0,0,1,1,1
229433,2024-04-18,NESTLEIND,-0.006924,-0.010091,0.040649,0,0,0,0,0,1
204788,2017-06-23,LTM,-0.052488,-0.052488,-0.010659,0,1,0,0,-1,0
54142,2011-11-04,BANKBARODA,-0.090892,-0.158156,0.004971,0,1,1,0,-1,1
332683,2018-12-04,ULTRACEMCO,-0.006091,-0.063042,0.013373,0,1,0,0,0,1
188073,2017-07-13,JINDALSTEL,-0.109426,-0.109426,0.144095,0,1,1,1,-1,1
280244,2022-11-23,SIEMENS,0.041412,-0.002166,0.096976,1,0,0,1,1,1


### Target distribution inside model-ready data

Now we should check the target distributions only inside model-ready rows, because that is what training will actually use.

In [7]:
model_target_mask = stock_targets["target_ready_v1_20d"] == 1
clean_model_target_mask = stock_targets["target_ready_v1_clean_20d"] == 1

target_label_cols = [
    "target_positive_return_20d",
    "target_downside_5pct_20d",
    "target_big_downside_10pct_20d",
    "target_upside_5pct_20d",
    "target_return_bucket_20d",
]

print("Model target rows:", model_target_mask.sum())
print("Clean model target rows:", clean_model_target_mask.sum())

for col in target_label_cols:
    print(f"\n{col} distribution inside target_ready_v1_20d:")

    counts = stock_targets.loc[model_target_mask, col].value_counts(dropna=False)
    pct = stock_targets.loc[model_target_mask, col].value_counts(normalize=True, dropna=False) * 100

    summary = pd.DataFrame(
        {
            "count": counts,
            "pct": pct.round(2),
        }
    )

    display(summary)

print("\nTarget-ready rows by stock acceptance status:")
display(
    stock_targets.loc[model_target_mask, "stock_acceptance_status"]
    .value_counts(dropna=False)
)

print("\nTarget-ready rows by year:")
display(
    stock_targets.loc[model_target_mask]
    .assign(year=lambda x: x["date"].dt.year)
    .groupby("year")
    .size()
    .reset_index(name="target_ready_rows")
)

Model target rows: 307453
Clean model target rows: 307418

target_positive_return_20d distribution inside target_ready_v1_20d:


,count,pct
target_positive_return_20d,,
1,174065,56.62
0,133388,43.38



target_downside_5pct_20d distribution inside target_ready_v1_20d:


,count,pct
target_downside_5pct_20d,,
0,190447,61.94
1,117006,38.06



target_big_downside_10pct_20d distribution inside target_ready_v1_20d:


,count,pct
target_big_downside_10pct_20d,,
0,263701,85.77
1,43752,14.23



target_upside_5pct_20d distribution inside target_ready_v1_20d:


,count,pct
target_upside_5pct_20d,,
0,154723,50.32
1,152730,49.68



target_return_bucket_20d distribution inside target_ready_v1_20d:


,count,pct
target_return_bucket_20d,,
1,127719,41.54
0,91614,29.8
-1,88120,28.66



Target-ready rows by stock acceptance status:


stock_acceptance_status
OK    307453
Name: count, dtype: int64


Target-ready rows by year:


,year,target_ready_rows
0,2011,8737
1,2012,17301
2,2013,19760
3,2014,19040
4,2015,19520
5,2016,19645
6,2017,20478
7,2018,20852
8,2019,21265
9,2020,22230


No <NA> inside target-ready rows.

Only OK stocks appear in target-ready rows.

Target classes are not badly imbalanced.

The year starts from 2011 because our features need enough history for 200-day / 252-day rolling windows. That is expected.

Now create a more MarketGuard-style scenario target.

This combines future downside and future upside.

### 20-day risk/reward scenario labels

The target asks two yes-or-no questions about the stock’s price movement during the next 20 trading days:

1. Did the stock fall by at least 10%?
2. Did the stock rise by at least 10%?

These two questions create four possible scenarios:

| ID | Hit downside threshold? | Hit upside threshold? | Scenario |
|---:|:---:|:---:|---|
| 0 | No | No | `quiet_or_rangebound` |
| 1 | No | Yes | `upside_low_drawdown` |
| 2 | Yes | Yes | `volatile_two_sided` |
| 3 | Yes | No | `downside_risk_limited_upside` |

In [8]:
# 20-day risk/reward scenario target.
#
# This uses future path behavior:
# - future_min_return_20d = worst drawdown seen within next 20 trading days
# - future_max_return_20d = best upside seen within next 20 trading days
#
# Scenario meanings:
# 0 = quiet_or_rangebound
# 1 = upside_low_drawdown
# 2 = volatile_two_sided
# 3 = downside_risk_limited_upside

scenario_available_mask = target_available_mask.copy()

stock_targets["target_scenario_20d"] = pd.Series(
    pd.NA,
    index=stock_targets.index,
    dtype="string"
)

stock_targets["target_scenario_id_20d"] = pd.Series(
    pd.NA,
    index=stock_targets.index,
    dtype="Int64"
)

future_min_20d = stock_targets["future_min_return_20d"]
future_max_20d = stock_targets["future_max_return_20d"]

quiet_or_rangebound_mask = (
    scenario_available_mask
    & (future_min_20d > DOWNSIDE_RISK_THRESHOLD_20D)
    & (future_max_20d < UPSIDE_THRESHOLD_20D)
)

upside_low_drawdown_mask = (
    scenario_available_mask
    & (future_min_20d > DOWNSIDE_RISK_THRESHOLD_20D)
    & (future_max_20d >= UPSIDE_THRESHOLD_20D)
)

volatile_two_sided_mask = (
    scenario_available_mask
    & (future_min_20d <= DOWNSIDE_RISK_THRESHOLD_20D)
    & (future_max_20d >= UPSIDE_THRESHOLD_20D)
)

downside_risk_limited_upside_mask = (
    scenario_available_mask
    & (future_min_20d <= DOWNSIDE_RISK_THRESHOLD_20D)
    & (future_max_20d < UPSIDE_THRESHOLD_20D)
)

stock_targets.loc[quiet_or_rangebound_mask, "target_scenario_20d"] = "quiet_or_rangebound"
stock_targets.loc[quiet_or_rangebound_mask, "target_scenario_id_20d"] = 0

stock_targets.loc[upside_low_drawdown_mask, "target_scenario_20d"] = "upside_low_drawdown"
stock_targets.loc[upside_low_drawdown_mask, "target_scenario_id_20d"] = 1

stock_targets.loc[volatile_two_sided_mask, "target_scenario_20d"] = "volatile_two_sided"
stock_targets.loc[volatile_two_sided_mask, "target_scenario_id_20d"] = 2

stock_targets.loc[downside_risk_limited_upside_mask, "target_scenario_20d"] = "downside_risk_limited_upside"
stock_targets.loc[downside_risk_limited_upside_mask, "target_scenario_id_20d"] = 3

# Realized future upside/downside ratio.
# This is a target-side diagnostic, not a feature.
future_downside_abs_20d = future_min_20d.clip(upper=0).abs()
future_upside_pos_20d = future_max_20d.clip(lower=0)

stock_targets["future_upside_downside_ratio_20d"] = np.where(
    scenario_available_mask & (future_downside_abs_20d > 0),
    future_upside_pos_20d / future_downside_abs_20d,
    np.nan
)

print("Stock targets shape:", stock_targets.shape)

print("\nScenario distribution, all available 20d targets:")
display(stock_targets["target_scenario_20d"].value_counts(dropna=False))

print("\nScenario distribution inside target_ready_v1_20d:")
display(
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        "target_scenario_20d"
    ].value_counts(dropna=False)
)

print("\nScenario ID distribution inside target_ready_v1_20d:")
display(
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        "target_scenario_id_20d"
    ].value_counts(dropna=False).sort_index()
)

display(
    stock_targets[
        [
            "date",
            "symbol",
            "future_return_20d",
            "future_min_return_20d",
            "future_max_return_20d",
            "future_upside_downside_ratio_20d",
            "target_scenario_20d",
            "target_scenario_id_20d",
            "target_ready_v1_20d",
        ]
    ]
    .dropna()
    .sample(30, random_state=42)
)

Stock targets shape: (357370, 176)

Scenario distribution, all available 20d targets:


target_scenario_20d
upside_low_drawdown             151796
downside_risk_limited_upside    110966
quiet_or_rangebound              67156
volatile_two_sided               25452
<NA>                              2000
Name: count, dtype: int64[pyarrow]


Scenario distribution inside target_ready_v1_20d:


target_scenario_20d
upside_low_drawdown             131151
downside_risk_limited_upside     95427
quiet_or_rangebound              59296
volatile_two_sided               21579
Name: count, dtype: int64[pyarrow]


Scenario ID distribution inside target_ready_v1_20d:


target_scenario_id_20d
0     59296
1    131151
2     21579
3     95427
Name: count, dtype: Int64

,date,symbol,future_return_20d,future_min_return_20d,future_max_return_20d,future_upside_downside_ratio_20d,target_scenario_20d,target_scenario_id_20d,target_ready_v1_20d
20664,2025-01-27,ADANIPOWER,-0.036728,-0.036728,0.052023,1.416442,upside_low_drawdown,1,1
218893,2013-04-17,MOTHERSON,0.158645,-0.025814,0.158645,6.145734,upside_low_drawdown,1,1
339072,2011-09-15,UNITDSPR,-0.011706,-0.142169,0.030817,0.216762,downside_risk_limited_upside,3,0
28184,2022-06-17,APOLLOHOSP,0.062263,-0.004796,0.063465,13.233502,upside_low_drawdown,1,1
346921,2017-04-19,VEDL,-0.012813,-0.052531,0.040145,0.764223,downside_risk_limited_upside,3,1
344913,2025-09-09,VBL,-0.083562,-0.083562,0.002740,0.032786,downside_risk_limited_upside,3,1
152198,2022-05-19,HINDALCO,-0.194005,-0.194005,0.029017,0.149568,downside_risk_limited_upside,3,1
272744,2025-07-31,SHREECEM,-0.030510,-0.049821,0.017202,0.345277,quiet_or_rangebound,0,1
330717,2010-12-09,ULTRACEMCO,-0.087529,-0.087529,0.013487,0.154091,downside_risk_limited_upside,3,0
176382,2025-03-28,INFY,-0.040588,-0.110718,-0.013052,0.000000,downside_risk_limited_upside,3,1


### Target-ready flags for all horizons

In [9]:
for horizon in TARGET_HORIZONS:
    target_ready_col = f"target_ready_v1_{horizon}d"
    target_ready_clean_col = f"target_ready_v1_clean_{horizon}d"

    available_mask = (
        (stock_targets[f"target_available_{horizon}d"] == 1)
        & (stock_targets[f"path_target_available_{horizon}d"] == 1)
    )

    stock_targets[target_ready_col] = (
        (stock_targets["model_ready_v1"] == 1)
        & available_mask
    ).astype(int)

    stock_targets[target_ready_clean_col] = (
        (stock_targets["model_ready_v1_clean"] == 1)
        & available_mask
    ).astype(int)

print("Stock targets shape:", stock_targets.shape)

for horizon in TARGET_HORIZONS:
    print(f"\nTarget-ready v1 {horizon}d:")
    display(stock_targets[f"target_ready_v1_{horizon}d"].value_counts())

    print(f"\nTarget-ready v1 clean {horizon}d:")
    display(stock_targets[f"target_ready_v1_clean_{horizon}d"].value_counts())

Stock targets shape: (357370, 180)

Target-ready v1 5d:


target_ready_v1_5d
1    308623
0     48747
Name: count, dtype: int64


Target-ready v1 clean 5d:


target_ready_v1_clean_5d
1    308588
0     48782
Name: count, dtype: int64


Target-ready v1 20d:


target_ready_v1_20d
1    307453
0     49917
Name: count, dtype: int64


Target-ready v1 clean 20d:


target_ready_v1_clean_20d
1    307418
0     49952
Name: count, dtype: int64


Target-ready v1 60d:


target_ready_v1_60d
1    304117
0     53253
Name: count, dtype: int64


Target-ready v1 clean 60d:


target_ready_v1_clean_60d
1    304083
0     53287
Name: count, dtype: int64

### Future relative-return targets

Did the stock beat Nifty 50 over the next 20 days?

Did the stock beat its sector over the next 20 days?

In [10]:
# Future relative-return targets:
# These are target-side values, not input features.
#
# Example:
# future_excess_return_20d_vs_nifty50 =
#     stock future 20d return - Nifty 50 future 20d return

for horizon in TARGET_HORIZONS:
    # Future Nifty 50 return from date T to T+H
    stock_targets[f"future_nifty50_close_{horizon}d"] = (
        stock_targets
        .groupby("yf_ticker")["nifty50_close"]
        .shift(-horizon)
    )

    stock_targets[f"future_nifty50_return_{horizon}d"] = (
        stock_targets[f"future_nifty50_close_{horizon}d"]
        / stock_targets["nifty50_close"]
        - 1
    )

    stock_targets[f"future_excess_return_{horizon}d_vs_nifty50"] = (
        stock_targets[f"future_return_{horizon}d"]
        - stock_targets[f"future_nifty50_return_{horizon}d"]
    )

    # Future sector proxy return from date T to T+H
    stock_targets[f"future_sector_close_{horizon}d"] = (
        stock_targets
        .groupby("yf_ticker")["sector_close"]
        .shift(-horizon)
    )

    stock_targets[f"future_sector_return_{horizon}d"] = (
        stock_targets[f"future_sector_close_{horizon}d"]
        / stock_targets["sector_close"]
        - 1
    )

    stock_targets[f"future_excess_return_{horizon}d_vs_sector"] = (
        stock_targets[f"future_return_{horizon}d"]
        - stock_targets[f"future_sector_return_{horizon}d"]
    )

# 20-day outperform labels
available_20d_mask = stock_targets["target_ready_v1_20d"] == 1

stock_targets["target_outperform_nifty50_20d"] = make_target_flag(
    stock_targets["future_excess_return_20d_vs_nifty50"] > 0,
    available_20d_mask
)

stock_targets["target_outperform_sector_20d"] = make_target_flag(
    stock_targets["future_excess_return_20d_vs_sector"] > 0,
    available_20d_mask
)

print("Stock targets shape:", stock_targets.shape)

print("\n20d outperform Nifty 50:")
display(stock_targets["target_outperform_nifty50_20d"].value_counts(dropna=False))

print("\n20d outperform sector:")
display(stock_targets["target_outperform_sector_20d"].value_counts(dropna=False))

print("\n20d outperform Nifty 50 inside target_ready_v1_20d:")
display(
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        "target_outperform_nifty50_20d"
    ].value_counts(dropna=False)
)

print("\n20d outperform sector inside target_ready_v1_20d:")
display(
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        "target_outperform_sector_20d"
    ].value_counts(dropna=False)
)

display(
    stock_targets[
        [
            "date",
            "symbol",
            "future_return_20d",
            "future_nifty50_return_20d",
            "future_sector_return_20d",
            "future_excess_return_20d_vs_nifty50",
            "future_excess_return_20d_vs_sector",
            "target_outperform_nifty50_20d",
            "target_outperform_sector_20d",
            "target_ready_v1_20d",
        ]
    ]
    .dropna()
    .sample(30, random_state=42)
)

Stock targets shape: (357370, 200)

20d outperform Nifty 50:


target_outperform_nifty50_20d
1       157658
0       149795
<NA>     49917
Name: count, dtype: Int64


20d outperform sector:


target_outperform_sector_20d
1       158018
0       149435
<NA>     49917
Name: count, dtype: Int64


20d outperform Nifty 50 inside target_ready_v1_20d:


target_outperform_nifty50_20d
1    157658
0    149795
Name: count, dtype: Int64


20d outperform sector inside target_ready_v1_20d:


target_outperform_sector_20d
1    158018
0    149435
Name: count, dtype: Int64

,date,symbol,future_return_20d,future_nifty50_return_20d,future_sector_return_20d,future_excess_return_20d_vs_nifty50,future_excess_return_20d_vs_sector,target_outperform_nifty50_20d,target_outperform_sector_20d,target_ready_v1_20d
3118,2022-08-22,ABB,0.097925,0.018613,0.024148,0.079312,0.073777,1,1,1
240384,2019-01-16,PFC,-0.057917,-0.013246,-0.018669,-0.044671,-0.039248,0,0,1
83033,2013-02-19,CGPOWER,-0.048427,-0.032619,-0.035694,-0.015808,-0.012733,0,0,1
273848,2013-07-03,SHRIRAMFIN,-0.070000,-0.005008,-0.118161,-0.064992,0.048161,0,1,1
95222,2013-10-04,COALINDIA,-0.044026,0.058546,0.052197,-0.102572,-0.096222,0,0,1
187158,2013-10-22,JINDALSTEL,-0.005462,-0.032848,-0.002234,0.027386,-0.003229,1,0,1
189118,2021-10-06,JINDALSTEL,0.033077,0.015346,0.012546,0.017731,0.020531,1,1,1
114936,2018-02-28,DRREDDY,-0.047000,-0.026785,-0.037878,-0.020215,-0.009121,0,0,1
194298,2023-04-21,JSWSTEEL,-0.008513,0.039171,0.021064,-0.047684,-0.029578,0,0,1
118272,2015-02-20,EICHERMOT,-0.027273,-0.032003,-0.031196,0.004729,0.003922,1,1,1


### Target quality summary

In [11]:
target_columns = [
    # Future prices / returns
    "future_adj_close_5d",
    "future_return_5d",
    "future_min_return_5d",
    "future_max_return_5d",
    "target_available_5d",
    "path_target_available_5d",
    "target_ready_v1_5d",
    "target_ready_v1_clean_5d",

    "future_adj_close_20d",
    "future_return_20d",
    "future_min_return_20d",
    "future_max_return_20d",
    "target_available_20d",
    "path_target_available_20d",
    "target_ready_v1_20d",
    "target_ready_v1_clean_20d",

    "future_adj_close_60d",
    "future_return_60d",
    "future_min_return_60d",
    "future_max_return_60d",
    "target_available_60d",
    "path_target_available_60d",
    "target_ready_v1_60d",
    "target_ready_v1_clean_60d",

    # 20d labels
    "target_positive_return_20d",
    "target_downside_5pct_20d",
    "target_big_downside_10pct_20d",
    "target_upside_5pct_20d",
    "target_return_bucket_20d",
    "target_scenario_20d",
    "target_scenario_id_20d",
    "future_upside_downside_ratio_20d",

    # Relative future targets
    "future_nifty50_return_20d",
    "future_sector_return_20d",
    "future_excess_return_20d_vs_nifty50",
    "future_excess_return_20d_vs_sector",
    "target_outperform_nifty50_20d",
    "target_outperform_sector_20d",
]

missing_target_cols = [
    col for col in target_columns
    if col not in stock_targets.columns
]

print("Missing target columns:", missing_target_cols)

if missing_target_cols:
    raise ValueError(f"Missing target columns: {missing_target_cols}")

target_quality = []

for col in target_columns:
    target_quality.append(
        {
            "column": col,
            "dtype": str(stock_targets[col].dtype),
            "missing_count": stock_targets[col].isna().sum(),
            "missing_pct": stock_targets[col].isna().mean() * 100,
            "unique_count": stock_targets[col].nunique(dropna=True),
        }
    )

target_quality = (
    pd.DataFrame(target_quality)
    .sort_values("missing_pct", ascending=False)
    .reset_index(drop=True)
)

display(target_quality)

target_quality.to_csv(TARGET_QUALITY_PATH, index=False)

print("Saved:", TARGET_QUALITY_PATH)

print("\nTarget-ready 20d row audit:")
print("Rows:", len(stock_targets))
print("target_ready_v1_20d:", stock_targets["target_ready_v1_20d"].sum())
print("target_ready_v1_clean_20d:", stock_targets["target_ready_v1_clean_20d"].sum())

print("\nMissing labels inside target_ready_v1_20d rows:")
label_cols_for_audit = [
    "target_positive_return_20d",
    "target_downside_5pct_20d",
    "target_big_downside_10pct_20d",
    "target_upside_5pct_20d",
    "target_return_bucket_20d",
    "target_scenario_20d",
    "target_scenario_id_20d",
    "target_outperform_nifty50_20d",
    "target_outperform_sector_20d",
]

display(
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        label_cols_for_audit
    ]
    .isna()
    .sum()
    .reset_index()
    .rename(columns={"index": "target_label", 0: "missing_count"})
)

Missing target columns: []


,column,dtype,missing_count,missing_pct,unique_count
0,future_upside_downside_ratio_20d,float64,55215,15.450374,261708
1,target_outperform_sector_20d,Int64,49917,13.967876,2
2,target_outperform_nifty50_20d,Int64,49917,13.967876,2
3,future_sector_return_20d,float64,19254,5.387693,38292
4,future_excess_return_20d_vs_sector,float64,19254,5.387693,338113
5,future_min_return_60d,float64,6000,1.678932,344331
6,future_return_60d,float64,6000,1.678932,351211
7,future_max_return_60d,float64,6000,1.678932,344755
8,future_adj_close_60d,float64,6000,1.678932,330130
9,future_nifty50_return_20d,float64,5978,1.672776,4073


Saved: E:\Projects\marketguard-india\data\interim\targets\target_quality_summary_v1.csv

Target-ready 20d row audit:
Rows: 357370
target_ready_v1_20d: 307453
target_ready_v1_clean_20d: 307418

Missing labels inside target_ready_v1_20d rows:


,target_label,missing_count
0,target_positive_return_20d,0
1,target_downside_5pct_20d,0
2,target_big_downside_10pct_20d,0
3,target_upside_5pct_20d,0
4,target_return_bucket_20d,0
5,target_scenario_20d,0
6,target_scenario_id_20d,0
7,target_outperform_nifty50_20d,0
8,target_outperform_sector_20d,0


### Target dictionary

In [12]:
target_dictionary_rows = []

for horizon in TARGET_HORIZONS:
    target_dictionary_rows.extend(
        [
            (
                f"future_adj_close_{horizon}d",
                f"Adjusted close price {horizon} trading days in the future",
            ),
            (
                f"future_return_{horizon}d",
                f"Final adjusted-close return after {horizon} trading days",
            ),
            (
                f"target_available_{horizon}d",
                f"1 if final {horizon}-day future return target is available",
            ),
            (
                f"future_min_adj_close_{horizon}d",
                f"Lowest adjusted close reached within next {horizon} trading days",
            ),
            (
                f"future_max_adj_close_{horizon}d",
                f"Highest adjusted close reached within next {horizon} trading days",
            ),
            (
                f"future_min_return_{horizon}d",
                f"Worst future return seen within next {horizon} trading days",
            ),
            (
                f"future_max_return_{horizon}d",
                f"Best future return seen within next {horizon} trading days",
            ),
            (
                f"path_target_available_{horizon}d",
                f"1 if full future {horizon}-day path is available",
            ),
            (
                f"target_ready_v1_{horizon}d",
                f"1 if row has v1 features and {horizon}-day target available",
            ),
            (
                f"target_ready_v1_clean_{horizon}d",
                f"1 if row has clean v1 features and {horizon}-day target available",
            ),
            (
                f"future_nifty50_close_{horizon}d",
                f"Nifty 50 close {horizon} trading days in the future",
            ),
            (
                f"future_nifty50_return_{horizon}d",
                f"Nifty 50 return after {horizon} trading days",
            ),
            (
                f"future_excess_return_{horizon}d_vs_nifty50",
                f"Stock future {horizon}-day return minus Nifty 50 future {horizon}-day return",
            ),
            (
                f"future_sector_close_{horizon}d",
                f"Mapped sector proxy close {horizon} trading days in the future",
            ),
            (
                f"future_sector_return_{horizon}d",
                f"Mapped sector proxy return after {horizon} trading days",
            ),
            (
                f"future_excess_return_{horizon}d_vs_sector",
                f"Stock future {horizon}-day return minus sector proxy future {horizon}-day return",
            ),
        ]
    )

target_dictionary_rows.extend(
    [
        (
            "target_positive_return_20d",
            "1 if final 20-day stock return is positive",
        ),
        (
            "target_downside_5pct_20d",
            "1 if stock falls at least 5% at any point within next 20 trading days",
        ),
        (
            "target_big_downside_10pct_20d",
            "1 if stock falls at least 10% at any point within next 20 trading days",
        ),
        (
            "target_upside_5pct_20d",
            "1 if stock rises at least 5% at any point within next 20 trading days",
        ),
        (
            "target_return_bucket_20d",
            "-1 weak, 0 neutral, 1 strong based on final 20-day return",
        ),
        (
            "target_scenario_20d",
            "20-day risk/reward scenario name",
        ),
        (
            "target_scenario_id_20d",
            "20-day scenario ID: 0 quiet, 1 upside, 2 volatile, 3 downside risk",
        ),
        (
            "future_upside_downside_ratio_20d",
            "Future 20-day upside divided by absolute downside",
        ),
        (
            "target_outperform_nifty50_20d",
            "1 if stock outperforms Nifty 50 over next 20 trading days",
        ),
        (
            "target_outperform_sector_20d",
            "1 if stock outperforms mapped sector proxy over next 20 trading days",
        ),
    ]
)

target_dictionary = pd.DataFrame(
    target_dictionary_rows,
    columns=["target", "description"]
)

target_dictionary["exists_in_dataset"] = target_dictionary["target"].isin(stock_targets.columns)

display(target_dictionary)

print("Missing target dictionary entries:")
display(target_dictionary[target_dictionary["exists_in_dataset"] == False])

target_dictionary.to_csv(TARGET_DICTIONARY_PATH, index=False)

print("Saved:", TARGET_DICTIONARY_PATH)

,target,description,exists_in_dataset
0,future_adj_close_5d,Adjusted close price 5 trading days in the future,True
1,future_return_5d,Final adjusted-close return after 5 trading days,True
2,target_available_5d,1 if final 5-day future return target is avail...,True
3,future_min_adj_close_5d,Lowest adjusted close reached within next 5 tr...,True
4,future_max_adj_close_5d,Highest adjusted close reached within next 5 t...,True
5,future_min_return_5d,Worst future return seen within next 5 trading...,True
6,future_max_return_5d,Best future return seen within next 5 trading ...,True
7,path_target_available_5d,1 if full future 5-day path is available,True
8,target_ready_v1_5d,1 if row has v1 features and 5-day target avai...,True
9,target_ready_v1_clean_5d,1 if row has clean v1 features and 5-day targe...,True


Missing target dictionary entries:


,target,description,exists_in_dataset


Saved: E:\Projects\marketguard-india\data\interim\targets\target_dictionary_v1.csv


### Final target table audit

In [13]:
# Final target dataset audit before saving

key_cols = ["yf_ticker", "date"]

duplicate_key_rows = stock_targets.duplicated(key_cols).sum()

numeric_cols = stock_targets.select_dtypes(include=["number"]).columns.tolist()

inf_summary = []

for col in numeric_cols:
    s = pd.to_numeric(stock_targets[col], errors="coerce")
    inf_count = np.isinf(s).sum()

    if inf_count > 0:
        inf_summary.append(
            {
                "column": col,
                "inf_count": inf_count,
            }
        )

inf_summary = pd.DataFrame(inf_summary)

# Check target availability consistency
availability_checks = []

for horizon in TARGET_HORIZONS:
    availability_checks.append(
        {
            "horizon": horizon,
            "target_available_count": stock_targets[f"target_available_{horizon}d"].sum(),
            "path_target_available_count": stock_targets[f"path_target_available_{horizon}d"].sum(),
            "target_ready_count": stock_targets[f"target_ready_v1_{horizon}d"].sum(),
            "target_ready_clean_count": stock_targets[f"target_ready_v1_clean_{horizon}d"].sum(),
        }
    )

availability_checks = pd.DataFrame(availability_checks)

# Check missing labels inside 20d target-ready rows
label_cols_for_audit = [
    "target_positive_return_20d",
    "target_downside_5pct_20d",
    "target_big_downside_10pct_20d",
    "target_upside_5pct_20d",
    "target_return_bucket_20d",
    "target_scenario_20d",
    "target_scenario_id_20d",
    "target_outperform_nifty50_20d",
    "target_outperform_sector_20d",
]

missing_labels_inside_ready = (
    stock_targets.loc[
        stock_targets["target_ready_v1_20d"] == 1,
        label_cols_for_audit
    ]
    .isna()
    .sum()
    .reset_index()
)

missing_labels_inside_ready.columns = ["target_label", "missing_count"]

print("Final target dataset shape:", stock_targets.shape)
print("Duplicate yf_ticker-date rows:", duplicate_key_rows)
print("Numeric columns checked:", len(numeric_cols))
print("Columns with infinite values:", len(inf_summary))
print("Latest date:", stock_targets["date"].max())
print("Unique stocks:", stock_targets["yf_ticker"].nunique())

print("\nAvailability checks:")
display(availability_checks)

print("\nInfinite value summary:")
display(inf_summary)

print("\nMissing labels inside target_ready_v1_20d rows:")
display(missing_labels_inside_ready)

Final target dataset shape: (357370, 200)
Duplicate yf_ticker-date rows: 0
Numeric columns checked: 184
Columns with infinite values: 0
Latest date: 2026-07-14 00:00:00
Unique stocks: 100

Availability checks:


,horizon,target_available_count,path_target_available_count,target_ready_count,target_ready_clean_count
0,5,356870,356870,308623,308588
1,20,355370,355370,307453,307418
2,60,351370,351370,304117,304083



Infinite value summary:


""



Missing labels inside target_ready_v1_20d rows:


,target_label,missing_count
0,target_positive_return_20d,0
1,target_downside_5pct_20d,0
2,target_big_downside_10pct_20d,0
3,target_upside_5pct_20d,0
4,target_return_bucket_20d,0
5,target_scenario_20d,0
6,target_scenario_id_20d,0
7,target_outperform_nifty50_20d,0
8,target_outperform_sector_20d,0


### Save final target dataset

In [14]:
# Sort final target table for consistent downstream use
stock_targets = (
    stock_targets
    .sort_values(["yf_ticker", "date"])
    .reset_index(drop=True)
)

# Save CSV version
stock_targets.to_csv(TARGET_OUTPUT_CSV, index=False)

# Save Parquet version
parquet_saved = False

try:
    stock_targets.to_parquet(TARGET_OUTPUT_PARQUET, index=False)
    parquet_saved = True
except Exception as e:
    print("Parquet save skipped or failed.")
    print("Reason:", repr(e))

print("Final target dataset shape:", stock_targets.shape)
print("Saved CSV:", TARGET_OUTPUT_CSV)

if parquet_saved:
    print("Saved Parquet:", TARGET_OUTPUT_PARQUET)

print("\nTarget-ready rows:")
for horizon in TARGET_HORIZONS:
    print(
        f"{horizon}d:",
        stock_targets[f"target_ready_v1_{horizon}d"].sum(),
        "| clean:",
        stock_targets[f"target_ready_v1_clean_{horizon}d"].sum(),
    )

print("\nLatest date:", stock_targets["date"].max())
print("Unique stocks:", stock_targets["yf_ticker"].nunique())

Final target dataset shape: (357370, 200)
Saved CSV: E:\Projects\marketguard-india\data\interim\targets\stock_features_with_targets_v1.csv
Saved Parquet: E:\Projects\marketguard-india\data\interim\targets\stock_features_with_targets_v1.parquet

Target-ready rows:
5d: 308623 | clean: 308588
20d: 307453 | clean: 307418
60d: 304117 | clean: 304083

Latest date: 2026-07-14 00:00:00
Unique stocks: 100


## Target Engineering Summary

This notebook created the first version of the MarketGuard India target dataset.

### Final outputs

- `data/interim/targets/stock_features_with_targets_v1.csv`
- `data/interim/targets/stock_features_with_targets_v1.parquet`
- `data/interim/targets/target_quality_summary_v1.csv`
- `data/interim/targets/target_dictionary_v1.csv`

### Final dataset shape

- Rows: **357,370**
- Columns: **200**
- Unique stocks: **100**
- Latest date: **2026-07-14**

### Target-ready rows

| Horizon | Target-ready rows | Clean target-ready rows |
|---|---:|---:|
| 5 trading days | 308,623 | 308,588 |
| 20 trading days | 307,453 | 307,418 |
| 60 trading days | 304,117 | 304,083 |

### Main 20-day scenario target

| Scenario ID | Scenario name | Rows inside target-ready data |
|---:|---|---:|
| 0 | quiet_or_rangebound | 59,296 |
| 1 | upside_low_drawdown | 131,151 |
| 2 | volatile_two_sided | 21,579 |
| 3 | downside_risk_limited_upside | 95,427 |

### Final validation checks

- Duplicate `yf_ticker` + `date` rows: **0**
- Infinite numeric values: **0**
- Missing labels inside `target_ready_v1_20d` rows: **0**

### Notes

Targets were created using future adjusted-close prices.

The dataset now supports:

- Future returns over 5, 20, and 60 trading days
- Future downside and upside path targets
- 20-day binary classification labels
- 20-day risk/reward scenario labels
- Future excess-return targets versus Nifty 50 and sector proxy

The latest rows do not have future targets because their future prices are not available yet. This is expected and handled using target availability flags.